qiskit/circuit/parametervector.py

```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Parameter Vector Class to simplify management of parameter lists.
+"""
+
+from .parameter import Parameter
+
+class ParameterVector():
+    """ParameterVector Class to quickly generate lists of parameters"""
+    def __init__(self, name, length=0):
+        self._name = name
+        self._params = []
+        for i in range(length):
+            self._params += [Parameter("{0}{1}".format(self._name, i))]
+
+    @property
+    def name(self):
+        """Returns the name of the ParameterVector."""
+        return self._name
+
+    @property
+    def params(self):
+        """Returns the list of parameters in the ParameterVector."""
+        return self._params
+
+    def __getitem__(self, offset):
+        return self.params[offset]
+
+    # Returns a list, not a ParameterVector
+    def __getslice__(self, low, high):
+        return self.params[low:high]
+
+    def __iter__(self):
+        return iter(self.params)
+
+    def __next__(self):
+        return next(self.params)
+
+    # Add a list of Parameter objects to the vector
+    # TODO maybe remove to make immutable?
+    def extend(self, param_list=None):
+        self._params += param_list
+
+    # Add a single Parameter object to the vector
+    # TODO maybe remove to make immutable?
+    def append(self, param=None):
+        self._params.append(param)
+
+    # Grow the vector to more elements
+    # TODO maybe remove to make immutable?
+    def grow(self, grow_length):
+        for i in range(grow_length):
+            self._params[i] = Parameter("{0}{1}".format(self._name, len(self)+i))
+
+    def __len__(self):
+        return len(self.params)
+
+    def __str__(self):
+        return '{}, {}'.format(self.name, [str(item) for item in self.params])
+
+    def __copy__(self):
```

Comment:
If `ParameterVector` is mutable, would we want this to be a normal copy? Say if I have a vector, copy it, and then extend the copy, do I want those changes to also appear in the original? (The way `Parameter` implements copy is sort of a hack that I'd like to revisit, but I don't think we need that behavior here.)

Reply:
I think the standard is for the array to be copied properly, but not the elements in the array - meaning if you added elements to your copy it wouldn't be reflected in the original, but if you changed an element object it would. Should we do that? If so, should I get rid of __deepcopy__?

Reply:
Yes, I think that's the behavior we want. With the `ParameterVector` immutable, this should be a non-issue, but unless there's a use case for the non-standard copy, I think we can remove `__copy__` and `__deepcopy__`.

Reply:
Makes sense, removing